ledger/complete/wal/checkpointer_test.go

```diff
@@ -443,57 +434,57 @@ func Test_Checkpointing(t *testing.T) {
-		require.NoError(t, err)
-
-		// create WAL
-
-		keys := testutils.RandomUniqueKeys(numInsPerStep, keyNumberOfParts, 1600, 1600)
-		values := testutils.RandomValues(numInsPerStep, valueMaxByteSize/2, valueMaxByteSize)
-		update, err := ledger.NewUpdate(ledger.State(trie.EmptyTrieRootHash()), keys, values)
-		require.NoError(t, err)
-
-		trieUpdate, err := pathfinder.UpdateToTrieUpdate(update, pathFinderVersion)
-		require.NoError(t, err)
-
-		_, _, err = wal.RecordUpdate(trieUpdate)
-		require.NoError(t, err)
-
-		// some buffer time of the checkpointer to run
-		time.Sleep(1 * time.Second)
-		<-wal.Done()
-
-		require.FileExists(t, path.Join(dir, "00000001")) //make sure WAL segment is saved
-
-		wal2, err := realWAL.NewDiskWAL(zerolog.Nop(), nil, metrics.NewNoopCollector(), dir, size*10, pathByteSize, segmentSize)
-		require.NoError(t, err)
-
-		checkpointer, err := wal2.NewCheckpointer()
-		require.NoError(t, err)
-
-		t.Run("write error", func(t *testing.T) {
-			errWrite := errors.New("unexpected write error")
-
-			err = checkpointer.Checkpoint(1, func() (io.WriteCloser, error) {
-				return newWriteCloserWithErrors(errWrite, nil), nil
-			})
-			require.ErrorIs(t, err, errWrite)
-		})
-
-		t.Run("close error", func(t *testing.T) {
-			errClose := errors.New("unexpected close error")
-
-			err = checkpointer.Checkpoint(1, func() (io.WriteCloser, error) {
-				return newWriteCloserWithErrors(nil, errClose), nil
-			})
-			require.ErrorIs(t, err, errClose)
-		})
-	})
-}
+// func TestCheckpointFileError(t *testing.T) {
+//
+// 	unittest.RunWithTempDir(t, func(dir string) {
+//
+// 		wal, err := realWAL.NewDiskWAL(zerolog.Nop(), nil, metrics.NewNoopCollector(), dir, size*10, pathByteSize, segmentSize)
+// 		require.NoError(t, err)
+//
+// 		// create WAL
+//
+// 		keys := testutils.RandomUniqueKeys(numInsPerStep, keyNumberOfParts, 1600, 1600)
+// 		values := testutils.RandomValues(numInsPerStep, valueMaxByteSize/2, valueMaxByteSize)
+// 		update, err := ledger.NewUpdate(ledger.State(trie.EmptyTrieRootHash()), keys, values)
+// 		require.NoError(t, err)
+//
+// 		trieUpdate, err := pathfinder.UpdateToTrieUpdate(update, pathFinderVersion)
+// 		require.NoError(t, err)
+//
+// 		_, _, err = wal.RecordUpdate(trieUpdate)
+// 		require.NoError(t, err)
+//
+// 		// some buffer time of the checkpointer to run
+// 		time.Sleep(1 * time.Second)
+// 		<-wal.Done()
+//
+// 		require.FileExists(t, path.Join(dir, "00000001")) //make sure WAL segment is saved
+//
+// 		wal2, err := realWAL.NewDiskWAL(zerolog.Nop(), nil, metrics.NewNoopCollector(), dir, size*10, pathByteSize, segmentSize)
+// 		require.NoError(t, err)
+//
+// 		checkpointer, err := wal2.NewCheckpointer()
+// 		require.NoError(t, err)
+//
+// 		t.Run("write error", func(t *testing.T) {
+// 			errWrite := errors.New("unexpected write error")
+//
+// 			err = checkpointer.Checkpoint(1, func() (io.WriteCloser, error) {
+// 				return newWriteCloserWithErrors(errWrite, nil), nil
+// 			})
```

Comment:
writer is now created internally, not sure how to re-create this case here, any idea?

Reply:
if you convert StoreCheckpoint into a method of checkpointer, you can store an extra field in checkpointer which normally points to CreateCheckpointWriterForFile, but can be override during test.

Reply:
Yeah, but I feel that's a trade off. 

If StoreCheckpoint is a function, then the behavior of storing a checkpoint is only determined by the StoreCheckpoint's inputs.
But if StoreCheckpoint is a method, then the behavior is not only determined by the inputs but also could possibly be altered by the object's state. If there is bug, the surface would be extended a lot, we have to check whether StoreCheckpoint internally depend on any state that could change, which makes the reasoning harder.